packages/firestore/src/model/proto_values.ts

```diff
@@ -0,0 +1,385 @@
+  const rightArray = right.values || [];
+
+  for (let i = 0; i < leftArray.length && i < rightArray.length; ++i) {
+    const valueCompare = compare(leftArray[i], rightArray[i]);
+    if (valueCompare) {
+      return valueCompare;
+    }
+  }
+  return primitiveComparator(leftArray.length, rightArray.length);
+}
+
+function compareMaps(left: api.MapValue, right: api.MapValue): number {
+  const leftMap = left.fields || {};
+  const leftKeys = keys(leftMap);
+  const rightMap = right.fields || {};
+  const rightKeys = keys(leftMap);
+
+  leftKeys.sort();
+  rightKeys.sort();
+
+  for (let i = 0; i < leftKeys.length && i < rightKeys.length; ++i) {
+    const keyCompare = primitiveComparator(leftKeys[i], rightKeys[i]);
+    if (keyCompare !== 0) {
+      return keyCompare;
+    }
+    const valueCompare = compare(leftMap[leftKeys[i]], rightMap[rightKeys[i]]);
+    if (valueCompare !== 0) {
+      return valueCompare;
+    }
+  }
+
+  return primitiveComparator(leftKeys.length, rightKeys.length);
+}
+
+/**
+ * Converts the possible Proto values for a timestamp value into a "seconds and
+ * nanos" representation.
+ */
+export function normalizeTimestamp(
+  date: string | { seconds?: string; nanos?: number }
+): { seconds: number; nanos: number } {
+  assert(!!date, 'Cannot deserialize null or undefined timestamp.');
+
+  // The json interface (for the browser) will return an iso timestamp string,
+  // while the proto js library (for node) will return a
+  // google.protobuf.Timestamp instance.
+  if (typeof date === 'string') {
+    // The date string can have higher precision (nanos) than the Date class
+    // (millis), so we do some custom parsing here.
+
+    // Parse the nanos right out of the string.
+    let nanos = 0;
+    const fraction = ISO_TIMESTAMP_REG_EXP.exec(date);
+    assert(!!fraction, 'invalid timestamp: ' + date);
+    if (fraction[1]) {
+      // Pad the fraction out to 9 digits (nanos).
+      let nanoStr = fraction[1];
+      nanoStr = (nanoStr + '000000000').substr(0, 9);
+      nanos = Number(nanoStr);
+    }
+
+    // Parse the date to get the seconds.
+    const parsedDate = new Date(date);
+    const seconds = Math.floor(parsedDate.getTime() / 1000);
+
+    return { seconds, nanos };
+  } else {
+    // TODO(b/37282237): Use strings for Proto3 timestamps
+    // assert(!this.options.useProto3Json,
+    //   'The timestamp instance format requires Proto JS.');
+    const seconds = normalizeNumber(date.seconds);
+    const nanos = normalizeNumber(date.nanos);
+    return { seconds, nanos };
+  }
+}
+
+/** Converts the possible Proto types for numbers into a JavaScript number. */
+export function normalizeNumber(value: number | string | undefined): number {
+  // TODO(bjornick): Handle int64 greater than 53 bits.
+  if (typeof value === 'number') {
+    return value;
+  } else if (typeof value === 'string') {
+    // Proto 3 uses the string values 'NaN' and 'Infinity'.
+    if (value === 'NaN') {
```

Comment:
`Number()` parses these values just fine. Do we really need to special case them?

Reply:
Good call. This is actually from the serializer, but we don't need to this code here (or there).